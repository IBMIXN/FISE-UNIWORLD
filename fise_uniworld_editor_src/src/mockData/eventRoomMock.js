const mockEventRoom = {
  title: "mockTitle",
  eventDate: "2021-03-17T00:00:00.000Z",
  meetingTables: [],
  background: "mockBackground",
  scene: "mockScene",
};

const mockEventRoomUpdate = {
  title: "newMockTitle",
  eventDate: "2021-02-27T12:00:00.000Z",
  meetingTables: [],
  background: "newMockBackground",
  scene: "newMockScene",
};

const mockMeetingTable = {
  title: "mockTitle",
  type: "RoundMeetingTable",
  logoUrl: "https://via.placeholder.com/150",
  zoomUrl: "https://zoom.us/",
  posX: 0,
  posY: 0,
};

export { mockEventRoom, mockEventRoomUpdate, mockMeetingTable };
