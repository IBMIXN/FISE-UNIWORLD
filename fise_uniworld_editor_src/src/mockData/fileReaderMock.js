export default class FileReaderMock {
  onerror = jest.fn();
  onload = jest.fn();
  addEventListener = (_, fn) => fn({ target: { result: {} } });
  readAsDataURL = jest.fn();
}
