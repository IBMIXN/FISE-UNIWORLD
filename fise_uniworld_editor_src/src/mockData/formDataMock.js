export default class FormDataMock {
  append = jest.fn();
}
