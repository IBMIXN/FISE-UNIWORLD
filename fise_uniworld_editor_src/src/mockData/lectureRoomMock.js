const mockLectureRoom = {
  title: "mockTitle",
  module: "mockModule",
  lecturer: "mockLecturer",
  startTime: "2021-03-17T09:00:00.000Z",
  endTime: "2021-03-17T10:00:00.000Z",
  firstSlideUrl: "mockUrl",
  numSlides: 1,
};

const mockLectureRoomUpdate = {
  title: "newMockTitle",
  module: "newMockModule",
  lecturer: "newMockLecturer",
  startTime: "2021-02-27T14:00:00.000Z",
  endTime: "2021-02-27T15:00:00.000Z",
  firstSlideUrl: "newMockUrl",
  numSlides: 3,
};

export { mockLectureRoom, mockLectureRoomUpdate };
