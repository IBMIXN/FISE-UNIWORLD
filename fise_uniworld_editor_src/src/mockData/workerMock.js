// https://github.com/webpack-contrib/worker-loader/issues/10#issuecomment-325977537
module.exports = Object.create(null);
