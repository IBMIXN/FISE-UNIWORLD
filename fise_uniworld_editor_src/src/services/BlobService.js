import { axiosConfig } from "../config";

class BlobService {
  url = "/blobs";

  async get(container) {
    return (await axiosConfig.get(`${this.url}/${container}`)).data.blobs;
  }

  async upload(data, container) {
    const config = {
      headers: {
        "content-type": "multipart/form-data",
      },
    };
    return (await axiosConfig.post(`${this.url}/${container}`, data, config)).data.url;
  }

  async delete(fileName, container) {
    const data = { fileName };
    return await axiosConfig.delete(`${this.url}/${container}`, { data });
  }

  async deleteSlides(fileName, numSlides) {
    const data = { fileName, numSlides };
    return await axiosConfig.delete(`${this.url}/slides`, { data });
  }
}

export default new BlobService();
