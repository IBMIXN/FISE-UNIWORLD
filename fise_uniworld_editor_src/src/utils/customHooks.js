import React from "react";

const useReducerState = (defaultState) => {
  const reducer = (state, newState) => ({ ...state, ...newState });
  return React.useReducer(reducer, defaultState);
};

export { useReducerState };
