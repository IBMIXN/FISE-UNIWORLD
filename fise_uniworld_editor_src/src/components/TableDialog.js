import PropTypes from "prop-types";
import { MeetingTablePropType } from "../propTypes/EventRoom";
import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import {
  Button,
  TextField,
  Dialog,
  DialogTitle,
  DialogActions,
  DialogContent,
  FormControl,
  FormControlLabel,
  Radio,
  RadioGroup,
  InputAdornment,
  Avatar,
  Typography,
  Grid,
} from "@material-ui/core";
import Autocomplete from "@material-ui/lab/Autocomplete";
import FileUpload from "./FileUpload";
import { useReducerState } from "../utils/customHooks";
import BlobService from "../services/BlobService";
import { getFileNameFromBlobUrl } from "../utils/blobs";

const useStyles = makeStyles((theme) => ({
  divider: {
    padding: theme.spacing(1),
  },
  textMuted: {
    color: "rgba(0, 0, 0, 0.54)",
  },
  padLeft: {
    paddingLeft: theme.spacing(1),
  },
  padRight: {
    paddingRight: theme.spacing(1),
  },
  flexDisplay: {
    display: "flex",
  },
}));

const TableDialog = ({ open, onClose, onSubmit, updateTable }) => {
  const classes = useStyles();
  const tableInitialState = {
    title: "",
    type: "RoundMeetingTable",
    zoomUrl: "",
    logoUrl: "",
  };
  const [table, setTable] = useReducerState(tableInitialState);
  const [logos, setLogos] = React.useState([]);

  React.useEffect(() => {
    if (updateTable) {
      setTable(updateTable);
    }
  }, [updateTable, setTable]);

  React.useEffect(() => {
    async function fetchLogos() {
      try {
        const fetchedLogos = await BlobService.get("logos");
        if (Array.isArray(fetchedLogos) && fetchedLogos.length > 0) {
          setLogos(fetchedLogos);
          setTable({ logoUrl: fetchedLogos[0] });
        }
      } catch (err) {
        console.log(err);
      }
    }
    fetchLogos();
  }, [setTable]);

  const resetTableState = () => {
    setTable(tableInitialState);
    if (Array.isArray(logos) && logos.length > 0) {
      setTable({ logoUrl: logos[0] });
    }
  };

  const handleClose = () => {
    resetTableState();
    onClose();
  };

  const handleUpload = async (e) => {
    e.preventDefault();
    const file = e.target.files[0];
    if (file && file.name.match(/\.(jpg|jpeg|png|gif)$/)) {
      const formData = new FormData();
      formData.append("image", e.target.files[0]);
      const logoUrl = await BlobService.upload(formData, "logos");
      setLogos([...logos, logoUrl]);
      setTable({ logoUrl });
    }
  };

  const handleSubmit = () => {
    table.logoUrl = table.logoUrl === "" ? undefined : table.logoUrl;
    table.zoomUrl = table.zoomUrl === "" ? undefined : table.zoomUrl;
    onSubmit(table);
    resetTableState();
    onClose();
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setTable({
      [name]: value,
    });
  };

  return (
    <Dialog id="dialog" open={open} onClose={handleClose} aria-labelledby="form-dialog-title">
      <DialogTitle id="form-dialog-title">
        {!!updateTable ? "Edit" : "Add"} a meeting table
      </DialogTitle>
      <DialogContent>
        <TextField
          autoFocus
          id="meetingTitle"
          margin="dense"
          name="title"
          label="Meeting Title"
          placeholder="e.g. Team 1 Project Demo"
          type="text"
          fullWidth
          value={table.title}
          onChange={handleChange}
          InputLabelProps={{
            shrink: true,
          }}
        />
        <div className={classes.divider} />
        <FormControl component="fieldset">
          <Typography variant="caption" className={classes.textMuted}>
            Table Type
          </Typography>
          <RadioGroup
            aria-label="type"
            id="tableType"
            name="type"
            value={table.type}
            onChange={handleChange}
            row
          >
            <FormControlLabel
              value="RoundMeetingTable"
              control={<Radio color="primary" />}
              label="Round Table"
            />
            <FormControlLabel
              value="RectangularMeetingTable"
              control={<Radio color="primary" />}
              label="Rectangular Table"
            />
          </RadioGroup>
        </FormControl>
        <TextField
          id="zoomUrl"
          margin="dense"
          name="zoomUrl"
          label="Zoom URL"
          placeholder="e.g. https://us04web.zoom.us/..."
          type="text"
          fullWidth
          value={table.zoomUrl}
          onChange={handleChange}
          InputLabelProps={{
            shrink: true,
          }}
        />
        <div className={classes.divider} />
        <Grid container>
          <Grid className={classes.flexDisplay} item xs>
            <Autocomplete
              id="selectLogo"
              className={classes.padRight}
              fullWidth
              disableClearable
              noOptionsText="No options. Please upload a logo."
              options={logos}
              getOptionLabel={(logo) => getFileNameFromBlobUrl(logo)}
              renderOption={(option) => (
                <React.Fragment>
                  <Avatar alt="logo" src={option || "https://via.placeholder.com/150"} />
                  <span className={classes.padLeft}>{getFileNameFromBlobUrl(option)}</span>
                </React.Fragment>
              )}
              value={table.logoUrl}
              onChange={(e, value) => {
                setTable({ logoUrl: value });
              }}
              renderInput={(params) => (
                <TextField
                  {...params}
                  label="Logo"
                  variant="outlined"
                  InputProps={{
                    ...params.InputProps,
                    startAdornment: (
                      <InputAdornment position="start">
                        {table.logoUrl && <Avatar alt="logo" src={table.logoUrl} />}
                      </InputAdornment>
                    ),
                  }}
                  InputLabelProps={{
                    shrink: true,
                  }}
                />
              )}
            />
            <FileUpload id="uploadLogo" name="uploadLogo" onChange={handleUpload} accept="image/*">
              Upload Logo
            </FileUpload>
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button id="closeButton" onClick={handleClose} color="primary">
          Cancel
        </Button>
        <Button id="submitButton" onClick={handleSubmit} color="primary">
          Save Meeting Table
        </Button>
      </DialogActions>
    </Dialog>
  );
};

TableDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  updateTable: MeetingTablePropType,
};

TableDialog.defaultProps = {
  updateTable: undefined,
};

export default TableDialog;
