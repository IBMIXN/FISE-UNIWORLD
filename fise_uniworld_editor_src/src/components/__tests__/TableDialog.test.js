import React from "react";
import { shallow } from "enzyme";
import TableDialog from "../TableDialog";
import { mockMeetingTable } from "../../mockData/eventRoomMock";
import BlobService from "../../services/BlobService";
import { useReducerState } from "../../utils/customHooks";

jest.mock("../../services/BlobService");
jest.mock("../../utils/customHooks", () => {
  return {
    useReducerState: jest.fn(),
  };
});

describe("TableDialog test", () => {
  const table = {
    title: "",
    type: "RoundMeetingTable",
    zoomUrl: "",
    logoUrl: "",
  };
  const setTable = jest.fn();

  const flushPromises = () => new Promise(setImmediate);
  beforeEach(() => {
    useReducerState.mockImplementation(() => [table, setTable]);
    jest.spyOn(console, "log").mockImplementation(() => {});
  });

  it("should match snapshot", () => {
    const component = shallow(
      <TableDialog
        open={false}
        onClose={() => {}}
        onSubmit={() => {}}
        updateTable={mockMeetingTable}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("should set table if updateTable prop exists", () => {
    shallow(
      <TableDialog
        open={false}
        onClose={() => {}}
        onSubmit={() => {}}
        updateTable={mockMeetingTable}
      />
    );
    expect(setTable).toBeCalledWith(mockMeetingTable);
  });

  it("should not set lectureRoom if updateLectureRoom prop is undefined", () => {
    shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    expect(setTable).not.toBeCalled();
  });

  it("should fetch logos from blob service on load", async () => {
    shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    await flushPromises();
    expect(BlobService.get).toBeCalledWith("logos");
  });

  it("should correctly set table logoUrl to fetched logo", async () => {
    BlobService.get.mockImplementationOnce(() => Promise.resolve(["mockLogoUrl"]));
    shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    await flushPromises();
    expect(BlobService.get).toBeCalledWith("logos");
    expect(setTable).toBeCalledWith({ logoUrl: "mockLogoUrl" });
  });

  it("should correctly handle meeting title text change event", () => {
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockEvent = { target: { name: "title", value: "mockMeetingTitle" } };
    component.find("#meetingTitle").first().simulate("change", mockEvent);
    expect(setTable).toBeCalledWith({ title: mockEvent.target.value });
  });

  it("should correctly handle table type change event", () => {
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockEvent = { target: { name: "type", value: "mockTableType" } };
    component.find("#tableType").first().simulate("change", mockEvent);
    expect(setTable).toBeCalledWith({ type: mockEvent.target.value });
  });

  it("should correctly handle zoom url change event", () => {
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockEvent = { target: { name: "zoomUrl", value: "mockUrl" } };
    component.find("#zoomUrl").first().simulate("change", mockEvent);
    expect(setTable).toBeCalledWith({ zoomUrl: mockEvent.target.value });
  });

  it("should correctly handle logo select autocomplete change event", () => {
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockLogo = "mockLogo";
    component.find("#selectLogo").first().simulate("change", null, mockLogo);
    expect(setTable).toBeCalledWith({ logoUrl: mockLogo });
  });

  it("should correctly upload logo", async () => {
    BlobService.upload.mockImplementationOnce(() => Promise.resolve("mockLogoUrl"));
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockEvent = { target: { files: [{ name: "mockLogo.jpg" }] }, preventDefault: () => {} };
    component.find("#uploadLogo").first().simulate("change", mockEvent);
    await flushPromises();
    expect(BlobService.upload.mock.calls[0][1]).toBe("logos");
    expect(setTable).toBeCalledWith({ logoUrl: "mockLogoUrl" });
  });

  it("should not upload logo if file format is not an image", async () => {
    const component = shallow(<TableDialog open={false} onClose={() => {}} onSubmit={() => {}} />);
    const mockEvent = {
      target: { files: [{ name: "invalidFile.txt" }] },
      preventDefault: () => {},
    };
    component.find("#uploadLogo").first().simulate("change", mockEvent);
    await flushPromises();
    expect(BlobService.upload).not.toBeCalled();
    expect(setTable).not.toBeCalled();
  });

  it("should correctly handle close dialog event", async () => {
    const mockClose = jest.fn();
    const component = shallow(<TableDialog open={false} onClose={mockClose} onSubmit={() => {}} />);
    component.find("#dialog").first().simulate("close");
    await flushPromises();
    expect(setTable).toBeCalledTimes(1);
    expect(mockClose).toBeCalled();
  });

  it("should correctly handle close button", async () => {
    const mockClose = jest.fn();
    const component = shallow(<TableDialog open={false} onClose={mockClose} onSubmit={() => {}} />);
    component.find("#closeButton").first().simulate("click");
    await flushPromises();
    expect(setTable).toBeCalledTimes(1);
    expect(mockClose).toBeCalled();
  });

  it("should correctly handle submit dialog event", async () => {
    useReducerState.mockImplementationOnce(() => [
      { ...table, type: "RectangularMeetingTable" },
      setTable,
    ]);
    const mockClose = jest.fn();
    const mockSubmit = jest.fn();
    const component = shallow(
      <TableDialog open={false} onClose={mockClose} onSubmit={mockSubmit} />
    );
    component.find("#submitButton").first().simulate("click");
    await flushPromises();
    expect(setTable).toBeCalledTimes(1);
    expect(mockSubmit).toBeCalledWith({
      ...table,
      type: "RectangularMeetingTable",
      logoUrl: undefined,
      zoomUrl: undefined,
    });
    expect(mockClose).toBeCalled();
  });
});
