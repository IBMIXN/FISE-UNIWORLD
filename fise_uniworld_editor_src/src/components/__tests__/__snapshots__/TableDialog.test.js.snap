// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDialog test should match snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  id="dialog"
  onClose={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    Edit
     a meeting table
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(TextField))
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      autoFocus={true}
      fullWidth={true}
      id="meetingTitle"
      label="Meeting Title"
      margin="dense"
      name="title"
      onChange={[Function]}
      placeholder="e.g. Team 1 Project Demo"
      type="text"
      value=""
    />
    <div
      className="makeStyles-divider-1"
    />
    <WithStyles(ForwardRef(FormControl))
      component="fieldset"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-textMuted-2"
        variant="caption"
      >
        Table Type
      </WithStyles(ForwardRef(Typography))>
      <ForwardRef(RadioGroup)
        aria-label="type"
        id="tableType"
        name="type"
        onChange={[Function]}
        row={true}
        value="RoundMeetingTable"
      >
        <WithStyles(ForwardRef(FormControlLabel))
          control={
            <WithStyles(ForwardRef(Radio))
              color="primary"
            />
          }
          label="Round Table"
          value="RoundMeetingTable"
        />
        <WithStyles(ForwardRef(FormControlLabel))
          control={
            <WithStyles(ForwardRef(Radio))
              color="primary"
            />
          }
          label="Rectangular Table"
          value="RectangularMeetingTable"
        />
      </ForwardRef(RadioGroup)>
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(TextField))
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      fullWidth={true}
      id="zoomUrl"
      label="Zoom URL"
      margin="dense"
      name="zoomUrl"
      onChange={[Function]}
      placeholder="e.g. https://us04web.zoom.us/..."
      type="text"
      value=""
    />
    <div
      className="makeStyles-divider-1"
    />
    <WithStyles(ForwardRef(Grid))
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-flexDisplay-5"
        item={true}
        xs={true}
      >
        <WithStyles(ForwardRef(Autocomplete))
          className="makeStyles-padRight-4"
          disableClearable={true}
          fullWidth={true}
          getOptionLabel={[Function]}
          id="selectLogo"
          noOptionsText="No options. Please upload a logo."
          onChange={[Function]}
          options={Array []}
          renderInput={[Function]}
          renderOption={[Function]}
          value=""
        />
        <FileUpload
          accept="image/*"
          id="uploadLogo"
          name="uploadLogo"
          onChange={[Function]}
        >
          Upload Logo
        </FileUpload>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      id="closeButton"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      id="submitButton"
      onClick={[Function]}
    >
      Save Meeting Table
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
